src/analyses/cfg_dominators.h

```diff
@@ -16,17 +16,219 @@ Author: Georg Weissenbacher, georg@weissenbacher.name
+  {
+  }
+
+  /// Create the dominators set of node_index
+  dominatorst(std::shared_ptr<datat> dominators_data, node_indext node_index)
+    : dominators_data(dominators_data),
+      node_index(node_index),
+      cached_distance(npos)
+  {
+  }
+
+  dominatorst(const dominatorst &other)
+    : dominators_data(other.dominators_data),
+      node_index(other.node_index),
+      cached_distance(other.cached_distance)
+  {
+  }
+
+  dominatorst &operator=(const dominatorst &rhs)
+  {
+    dominators_data = rhs.dominators_data;
+    node_index = rhs.node_index;
+    cached_distance = rhs.cached_distance;
+    return *this;
+  }
+
+  class dominatorst_iteratort
+    : public std::iterator<std::forward_iterator_tag, T>
+  {
+  public:
+    using parentt = const datat *;
+    using elemt = const T;
+
+  private:
+    parentt data;
+    node_indext current_index;
+
+  public:
+    dominatorst_iteratort(parentt cfg_dominators, node_indext current_index)
+      : data(cfg_dominators), current_index(current_index)
+    {
+    }
+
+    dominatorst_iteratort &operator++()
+    {
+      INVARIANT(
+        current_index != npos, "Shouldn't try to increment end-iterator");
+      current_index = data->immediate_dominator[current_index];
+      return *this;
+    }
+
+    dominatorst_iteratort operator++(int)
+    {
+      INVARIANT(
+        current_index != npos, "Shouldn't try to post-increment end-iterator");
+      node_indext tmp = current_index;
+      ++(*this);
+      return dominatorst_iteratort(data, tmp);
+    }
+
+    const elemt *get() const
+    {
+      INVARIANT(
+        current_index != npos, "Shouldn't try to dereference end-iterator");
+      return &data->data[current_index];
+    }
+
+    const elemt &operator*() const
+    {
+      return *get();
+    }
+
+    const elemt *operator->() const
+    {
+      return get();
+    }
+
+    bool operator!=(const dominatorst_iteratort &other) const
+    {
+      INVARIANT(
+        data == other.data, "iterators from different sets are not comparable");
+      return current_index != other.current_index;
+    }
+
+    bool operator==(const dominatorst_iteratort &other) const
+    {
+      return !(*this != other);
+    }
+  };
+
+  using const_iterator = dominatorst_iteratort;
+
+  const_iterator begin() const
+  {
+    return const_iterator(dominators_data.get(), node_index);
+  }
+
+  const_iterator cbegin() const
+  {
+    return begin();
+  }
+
+  const_iterator end() const
+  {
+    return const_iterator(dominators_data.get(), npos);
+  }
+
+  const_iterator cend() const
+  {
+    return end();
+  }
+
+  /// Return an iterator node if node is in this dominators set, end() otherwise
+  /// Note: O(n), when making many queries against the same set it is probably
+  /// worth copying into a std::set
+  const_iterator find(const T &node) const
+  {
+    std::less<T> less;
+    // FIXME This works around a bug in other parts of the code in particular,
```

Comment:
What raised my eyebrows was the part below: "in particular, dependence_graph.cpp, where iterators to different lists than those that
are stored in this set are passed to find." That sounds like a problem that's avoidable, as opposed to what you are saying above, which seems inherent in the design of the data structures?

Reply:
If I'm understanding the dependence_graph.cpp correctly, then this 'property' of comparing items from different iterators is a fairly fundamental aspect of how dependence_graph works with goto programs, in that it is keeping around sets of `const_targett` where they may be from different goto-functions (and hence, different iterators) in the case of cross-function dependencies. It seems that dependence_graph.cpp already uses a similar pattern to @hannes-steffenhagen-diffblue in various places, e.g. `dep_graph_domaint::merge` is using `<` for comparisons when merging dependencies (OK, so that's partly just the merge algorithm but also means it tolerates `const_targett`s in different functions/iterators).

I think any attempt to fixup dependence_graph.cpp would definitely be outside the scope of this particular PR, but we can always raise an issue to cover that and come back to that in a different PR if people feel strongly enough that it should be fixed (of course, any out-of-tree code that builds on dependence_graph.cpp might end up also having to be fixed up in that case too).  As it is, I think this PR is probably at least in keeping with the current implementation of dependence_graph.cpp.

However, as @martin-cs has mentioned, it's probably worth holding off merging this though until we find out if our end user gets any benefit in their use-case.